urld499b6e5aa_HatlessFox_StPatrickSimulation_tgz-pJ8-ru_spbau_sd_MainJ8/src/ru/spbau/sd/model/game/PoliceStation.java

```diff
@@ -0,0 +1,143 @@
+  
+  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+  OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+  IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+  DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+  TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
+  OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+package ru.spbau.sd.model.game;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import ru.spbau.sd.model.framework.EndTurnListener;
+import ru.spbau.sd.model.framework.Field;
+import ru.spbau.sd.model.framework.FieldObject;
+import ru.spbau.sd.model.framework.Generator;
+import ru.spbau.sd.model.framework.Point2D;
+
+public class PoliceStation extends Generator implements EndTurnListener {
+
+    private List<Policeman> mWaitingOfficers = new ArrayList<>();
+    private List<Policeman> mWalkingOfficers = new ArrayList<>();
+    private Set<Drinker> mDrinkersToBeCaught = new HashSet<>();
+    
+    private Policeman mJustReturnedOfficer;
+    
+    public PoliceStation(int x, int y, int officersCnt) {
+        super(x, y);
+        
+        for (int i = 0; i < officersCnt; i++) {
+            mWaitingOfficers.add(new Policeman(-1, -1, this));
+        }
+        
+        Field.getInstance().addEndTurnListener(this);
+    }
+    
+    public List<Policeman> getWalkingOfficers() {
+        return Collections.unmodifiableList(mWalkingOfficers);
+    }
+    
+    private void processReturnedOfficers() {
+        for (Policeman p : mWalkingOfficers) {
+            if (!getEntryPoints().contains(p.getPosition())) { continue; }
+            if (!p.isGoingHome()) { break; }
+            
+            mDrinkersToBeCaught.remove(p.getCatchedBadGuy());
+            //remove policeman and drinker from the field
+            if (p.isBadGuyCaught()) {
+                Field.getInstance().removeMovable(p.getCatchedBadGuy());
+            }
+            Field.getInstance().removeMovable(p);
+            
+            //policeman is waiting for next drinker
+            mWalkingOfficers.remove(p);
+            mJustReturnedOfficer = p;
+            break;
+        }
+    }
+    
+    private boolean drinkerShouldBeCaught(Drinker d) {
+        if (mDrinkersToBeCaught.contains(d)) { return false; }
+        return d.getMovementStrategy() == Drinker.MovementStrategy.LAYING ||
+               d.getMovementStrategy() == Drinker.MovementStrategy.SLEEP;
+    }
+    
+    private Point2D drinkerIsReachable(Drinker d) {
+        for (Point2D ep : getEntryPoints()) {
+            if (!ep.equals(GameUtils.lookUpNextStep(ep, d.getPosition()))) {
+                return ep;
+            }
+        }
+        return null;
+    }
+    
+    private void processUncaughtDrinkers() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 94-124 into a separate method.